src/compiler/checker.ts

```diff
@@ -12495,6 +12509,36 @@ namespace ts {
         }
 
         /**
+         * Checks that the return type provided is an instantiation of the global Promise<T> type

```

Comment:
The current support we have in TypeScript 1.7.x for async functions allows you to specify a custom Promise implementation in the return type annotation. This was intended to support runtimes that do not yet define a global Promise implementation, and was primarily intended to support down-level emit for async functions in ES5, which is a goal for TypeScript 2.0.  This allows you to write an async function in the following fashion:

```ts
import * as bluebird from 'bluebird';

export async function fn(): bluebird.Promise {
}

console.log(fn() instanceof bluebird.Promise); // true 
```

However, this approach is not compatible with the ES7 approach to async functions which only leverages the native Promise implementation. As this native Promise is specified as part of ES6, we have elected to only allow you to supply a global `Promise<T>` return type for ES6, but reserve the previous functionality to eventually support async functions down-level.

If we were to then allow you to specify the return type of an async function as any type to which `Promise<T>` is assignable, then the above sample would still compile successfully, due to how "bluebird.d.ts" is currently implemented on DefinitelyTyped. The result is that upgrading from TypeScript 1.7.x to TypeScript 1.8 would result in completely different runtime semantics with *no indication to the developer* that anything had changed.

```ts
import * as bluebird from 'bluebird';

export async function fn(): bluebird.Promise {
}

console.log(fn() instanceof bluebird.Promise); // false?! 
```

By restricting the return type annotation to **only** the global `Promise<T>` type, we have the ability to loosen this restriction in the future. Yes, this does not fully align with how we support type annotations on generator functions; however, this is the only reliable way forward. We may, in a future release, opt to add a different mechanism for defining a compatible Promise implementation for ES5 async functions. If that does indeed become the case, we would be able to loosen the return type restriction. 

As this is a breaking change from TypeScript 1.7 behavior, it also is best to be more conservative for one to two releases to ensure any existing implementations have adjusted to this change before examining the feasibility of relaxing this restriction.